pywhy_stats/kernel_utils.py

```diff
@@ -0,0 +1,281 @@
+    """Attempt to infer the kernel function from the data type of X.
+
+    Parameters
+    ----------
+    X : ArrayLike
+        Data array. If the data type is not string, the RBF kernel is used.
+        If data type is string, the delta kernel is used.
+
+    Returns
+    -------
+    Callable[[ArrayLike], ArrayLike]
+        The kernel function.
+    """
+    if X.dtype.type not in (np.str_, np.object_):
+        return "rbf", {"gamma": 0.5 * estimate_squared_sigma_rbf(X)}
+    else:
+        return "delta", dict()
+
+
+def _normalize_data(X: ArrayLike) -> ArrayLike:
+    """Normalize data to zero mean and unit variance."""
+    for column in range(X.shape[1]):
+        if isinstance(X[0, column], int) or isinstance(X[0, column], float):
+            X[:, column] = stats.zscore(X[:, column])
+            X[:, column] = np.nan_to_num(X[:, column])  # in case some dim of X is constant
+    return X
+
+
+def compute_kernel(
+    X: ArrayLike,
+    Y: Optional[ArrayLike] = None,
+    metric: Optional[Union[Callable, str]] = None,
+    centered: bool = True,
+    n_jobs: Optional[int] = None,
+) -> Tuple[ArrayLike, float]:
+    """Compute a kernel matrix and corresponding width.
+
+    Also optionally estimates the kernel width parameter.
+
+    Parameters
+    ----------
+    X : ArrayLike of shape (n_samples_X, n_features_X)
+        The X array.
+    Y : ArrayLike of shape (n_samples_Y, n_features_Y), optional
+        The Y array, by default None.
+    metric : str, optional
+        The metric to compute the kernel function, by default 'rbf'.
+        Can be any string as defined in
+        :func:`sklearn.metrics.pairwise.pairwise_kernels`. Note 'rbf'
+        and 'gaussian' are the same metric.
+    centered : bool, optional
+        Whether to center the kernel matrix or not, by default True.
+        When centered, the kernel matrix induces a zero mean. The main purpose of
+        centering is to remove the bias or mean shift from the data represented
+        by the kernel matrix.
+    n_jobs : int, optional
+        The number of jobs to run computations in parallel, by default None.
+
+    Returns
+    -------
+    kernel : ArrayLike of shape (n_samples_X, n_samples_X) or (n_samples_X, n_samples_Y)
+        The kernel matrix.
+    """
+    # if the width of the kernel is not set, then use the median trick to set the
+    # kernel width based on the data X
+    if metric is None:
+        metric, kernel_params = _get_default_kernel(X)
+    else:
+        kernel_params = dict()
+
+    # compute the potentially pairwise kernel matrix
+    # If the number of arguments is just one, then we bypass the pairwise kernel
+    # optimized computation via sklearn and opt to use the metric function directly
+    if callable(metric) and len(inspect.getfullargspec(metric).args) == 1:
+        kernel = metric(X)
+    else:
+        kernel = pairwise_kernels(
+            X, Y=Y, metric=metric, n_jobs=n_jobs, filter_params=False, **kernel_params
```

Comment:
What if the `metric` is already providing a kernel matrix? This is, what if a user provides a custom kernel implementation instead of a metric? Currently, it is confusing that we ask for a `kernel_X` etc., but it is actually supposed to be a `metric` or do I miss something?

Reply:
The `if` statement above is meant to handle this and allow a user to provide a custom kernel implementation. So my take is we should have easy support for existing optimized kernels: e.g. scikit-learn/scipy strings. Therefore `kernel_X` can both be a callable that is documented [here](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.pairwise_kernels.html), or a string that we vectorize.

I agree we should also support custom kernels, but then we want the code to be consistent. In the previous implementation, the API inherently assumed it was an input of just `X`. But then it doesn't support pairwise kernels among X and Y. I think an easy way to handle this is to require the user to define a lambda that takes in `X, Y=None` and if Y is not used, then it is not used, but then we can have a consistent API signature.

I think the current way I handled it is not ideal yet, but it's just a draft to get us to converge on something that generalizes to the rest of the codebase.

Reply:
Ah ok, I thought it would need to be a metric then (not a kernel). In that way, it is basically similar to my suggestion above with the if-else. I see the point with `Y = None`, however, I would simply check if Y is None and if not, pass it as second argument. It will then either raise an error or the user provided a kernel that accepts a second argument. No need to add additional validation logic.

Reply:
I refactored it to use your suggestion. Is this what you meant?

Reply:
See lines above this. I.e.

```
    if callable(metric):
        # If the number of arguments is just one, then we bypass the pairwise kernel
        # optimized computation via sklearn and opt to use the metric function directly
        input_args = [X]
        if Y is not None:
            input_args.append(Y)
        kernel = metric(*input_args)
```